pkg/kv/kvserver/apply/asserter.go

```diff
@@ -0,0 +1,92 @@
+// Copyright 2023 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+// TODO: put in a test package
+package apply
+
+import (
+	fmt "fmt"
+
+	"github.com/cockroachdb/cockroach/pkg/kv/kvpb"
+	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/kvserverbase"
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
+	"go.etcd.io/raft/v3/raftpb"
+)
+
+type appliedCmd struct {
+	cmdID    kvserverbase.CmdIDKey
+	rangeID  roachpb.RangeID
+	logIndex kvpb.RaftIndex
+	replicas map[roachpb.ReplicaID]bool
+}
+
+// Asserter is a test utility that tracks application of Raft
+// commands, and asserts that a command is never applied at different log
+// indexes (i.e. a double-apply or replay).
+type Asserter struct {
+	mu syncutil.Mutex
+	// appliedCmds tracks applied commands by range and command ID.
+	appliedCmds map[roachpb.RangeID]map[kvserverbase.CmdIDKey]*appliedCmd
+}
+
+// NewAsserter creates a new asserter.
+func NewAsserter() *Asserter {
+	return &Asserter{
+		appliedCmds: map[roachpb.RangeID]map[kvserverbase.CmdIDKey]*appliedCmd{},
+	}
+}
+
+// Apply records a command being applied, and asserts validity. It is safe
+// to call multiple times for the same command, as is commonly done via
+// StoreTestingKnobs.TestingApplyCalledTwiceFilter.
+//
+// TODO: we need to handle snapshots too, by treating all commands up to the
+// snapshot index as applied on that replica.
```

Comment:
The beauty of the approach here is the property: for every command that applies, at least one replica calls `Apply`. It's totally possible that N-1 other replicas will receive a snapshot and won't report this command, but there is always one who does. (consider documenting this)

So, in the end `appliedCmds[rangeID]` contains all the commands the range applied. We can also reconstruct the log of all applied commands if we sort it by index/LAI.

Reply:
Yep. Will add a comment when I add snapshothandling. We probably also want to track and assert entry commits as well, but this will do for now.